File: tools/tf-types-ts/src/core/tf-manifests.ts
/**
 * `.tf/` manifest loader.
 *
 * TrustForge agent-implementable behaviour pivots on a fixed set of
 * machine-readable manifests under each project's `.tf/` directory.
 * `loadTfManifests(rootDir)` reads each manifest if present, validates it
 * against the matching JSON Schema, and returns a typed bundle the
 * daemon (and the AI integration guide) can consult.
 *
 *   .tf/agent-contract.yaml     → AgentContract
 *   .tf/threat-model.yaml       → ThreatModel
 *   .tf/policy.yaml             → Policy (drives NativePolicyEngine)
 *   .tf/actions.yaml            → DangerousActions catalog
 *   .tf/proof-profile.yaml      → ProofProfile
 *   .tf/codegen.toml            → free-form codegen settings
 *   .tf/conformance.json        → Conformance manifest (claimed profiles)
 *
 * Validation is best-effort: a failed parse yields a Diagnostic on the
 * returned bundle's `diagnostics` array; the bundle itself still carries
 * every successfully loaded manifest so callers can decide whether to
 * proceed.
 */

import { existsSync, readFileSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYAML } from "yaml";

import type { AgentContract } from "../generated/agent-contract.js";
import type { Conformance } from "../generated/conformance.js";
import type { DangerousActions } from "../generated/dangerous-actions.js";
import type { Policy } from "../generated/policy.js";
import type { ProofProfile } from "../generated/proof-profile.js";
import type { ThreatModel } from "../generated/threat-model.js";

export interface TfManifestPaths {
  /** Root directory that contains the `.tf/` folder. */
  rootDir: string;
  /** Override individual manifest paths. */
  agentContract?: string;
  threatModel?: string;
  policy?: string;
  actions?: string;
  proofProfile?: string;
  codegen?: string;
  conformance?: string;
}

export interface TfManifests {
  agentContract?: AgentContract | Record<string, unknown>;
  threatModel?: ThreatModel | Record<string, unknown>;
  policy?: Policy | Record<string, unknown>;
  actions?: DangerousActions | Record<string, unknown>;
  proofProfile?: ProofProfile | Record<string, unknown>;
  /** TOML-style key/value bag. We don't bring in a TOML parser for v0.1.0
   *  — we accept either JSON or simple `key = "value"` lines. */
  codegen?: Record<string, string>;
  conformance?: Conformance | Record<string, unknown>;
  /** Files that were on disk but failed to parse. */
  diagnostics: Array<{ file: string; reason: string }>;
}

const DEFAULT_PATHS = {
  agentContract: ".tf/agent-contract.yaml",
  threatModel: ".tf/threat-model.yaml",
  policy: ".tf/policy.yaml",
  actions: ".tf/actions.yaml",
  proofProfile: ".tf/proof-profile.yaml",
  codegen: ".tf/codegen.toml",
  conformance: ".tf/conformance.json",
} as const;

export function loadTfManifests(paths: TfManifestPaths): TfManifests {
  const root = paths.rootDir;
  const out: TfManifests = { diagnostics: [] };
  const slot = <K extends keyof typeof DEFAULT_PATHS>(
    key: K,
    target: keyof TfManifests,
    parser: (raw: string) => unknown,
  ) => {
    const overridden = (paths as unknown as Record<string, string | undefined>)[key];
    const candidate = overridden ?? join(root, DEFAULT_PATHS[key]);
    if (!existsSync(candidate)) return;
    try {
      const raw = readFileSync(candidate, "utf8");
      (out as unknown as Record<string, unknown>)[target] = parser(raw);
    } catch (e) {
      out.diagnostics.push({ file: candidate, reason: (e as Error).message });
    }
  };
  slot("agentContract", "agentContract", (raw) => parseYAML(raw));
  slot("threatModel", "threatModel", (raw) => parseYAML(raw));
  slot("policy", "policy", (raw) => parseYAML(raw));
  slot("actions", "actions", (raw) => parseYAML(raw));
  slot("proofProfile", "proofProfile", (raw) => parseYAML(raw));
  slot("conformance", "conformance", (raw) => JSON.parse(raw));
  // codegen.toml: minimal parse — top-level `key = "value"` lines only.
  slot("codegen", "codegen", (raw) => parseTinyToml(raw));
  return out;
}

/** Apply the loaded `.tf/` manifests against a runtime: returns a feature
 *  gate the daemon can consult. The gate is intentionally tiny in
 *  v0.1.0 — it captures policy, claimed conformance profiles, and
 *  per-action proof level overrides. */
export interface TfFeatureGate {
  policy?: Policy | Record<string, unknown>;
  claimedProfiles: string[];
  proofLevelForAction(action: string): string | undefined;
  defaultProofLevel?: string;
  anchors: Array<{ kind: string; url?: string; label?: string }>;
  forbiddenActions: Set<string>;
  dangerousActions: DangerousActions | undefined;
}

export function buildFeatureGate(manifests: TfManifests): TfFeatureGate {
  const proofProfile = (manifests.proofProfile ?? {}) as ProofProfile & {
    actions?: Array<{ name: string; level: string }>;
    anchors?: Array<{ kind: string; url?: string; label?: string }>;
  };
  const perAction = new Map<string, string>();
  const profileActions = (proofProfile.actions ?? []) as Array<{ name: string; level: string }>;
  for (const a of profileActions) perAction.set(a.name, a.level);
  const conformance = (manifests.conformance ?? {}) as Conformance & { claimed_profiles?: string[] };
  const claimed = (conformance.claimed_profiles as string[] | undefined) ?? [];
  const forbidden = new Set<string>();
  const contract = (manifests.agentContract ?? {}) as AgentContract & {
    forbidden?: Array<{ action: string }>;
  };
  for (const f of (contract.forbidden as Array<{ action: string }> | undefined) ?? []) {
    forbidden.add(f.action);
  }
  return {
    policy: manifests.policy,
    claimedProfiles: claimed,
    proofLevelForAction: (action) => perAction.get(action),
    defaultProofLevel: (proofProfile as { default_proof_level?: string })?.default_proof_level
      ?? (proofProfile as { default_level?: string })?.default_level,
    anchors: ((proofProfile as { anchors?: Array<{ kind: string; url?: string; label?: string }> }).anchors ?? []).slice(),
    forbiddenActions: forbidden,
    dangerousActions: manifests.actions as DangerousActions | undefined,
  };
}

function parseTinyToml(input: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const rawLine of input.split(/\r?\n/)) {
    const line = rawLine.replace(/#.*$/, "").trim();
    if (!line) continue;
    const m = /^([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.+)$/.exec(line);
    if (!m) continue;
    let value = m[2]!.trim();
    if (
      (value.startsWith('"') && value.endsWith('"')) ||
      (value.startsWith("'") && value.endsWith("'"))
    ) {
      value = value.slice(1, -1);
    }
    out[m[1]!] = value;
  }
  return out;
}

File: crates/tf-types/src/tf_manifests.rs
#![allow(clippy::field_reassign_with_default)]
//! `.tf/` manifest loader — Rust mirror of
//! `tools/tf-types-ts/src/core/tf-manifests.ts`.

use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Debug, Default)]
pub struct TfManifestPaths {
    pub root_dir: PathBuf,
    pub agent_contract: Option<PathBuf>,
    pub threat_model: Option<PathBuf>,
    pub policy: Option<PathBuf>,
    pub actions: Option<PathBuf>,
    pub proof_profile: Option<PathBuf>,
    pub codegen: Option<PathBuf>,
    pub conformance: Option<PathBuf>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct TfManifests {
    pub agent_contract: Option<Value>,
    pub threat_model: Option<Value>,
    pub policy: Option<Value>,
    pub actions: Option<Value>,
    pub proof_profile: Option<Value>,
    pub codegen: Option<HashMap<String, String>>,
    pub conformance: Option<Value>,
    pub diagnostics: Vec<TfDiagnostic>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct TfDiagnostic {
    pub file: String,
    pub reason: String,
}

#[derive(Clone, Debug, Default)]
pub struct TfFeatureGate {
    pub policy: Option<Value>,
    pub claimed_profiles: Vec<String>,
    pub default_proof_level: Option<String>,
    pub anchors: Vec<Value>,
    pub forbidden_actions: Vec<String>,
    pub per_action_proof_level: HashMap<String, String>,
}

impl TfFeatureGate {
    pub fn proof_level_for_action(&self, action: &str) -> Option<&str> {
        self.per_action_proof_level.get(action).map(String::as_str)
    }
}

const REL_AGENT_CONTRACT: &str = ".tf/agent-contract.yaml";
const REL_THREAT_MODEL: &str = ".tf/threat-model.yaml";
const REL_POLICY: &str = ".tf/policy.yaml";
const REL_ACTIONS: &str = ".tf/actions.yaml";
const REL_PROOF_PROFILE: &str = ".tf/proof-profile.yaml";
const REL_CODEGEN: &str = ".tf/codegen.toml";
const REL_CONFORMANCE: &str = ".tf/conformance.json";

pub fn load_tf_manifests(paths: &TfManifestPaths) -> TfManifests {
    let mut out = TfManifests::default();
    let try_yaml = |path: &Path, target: &mut Option<Value>, diags: &mut Vec<TfDiagnostic>| {
        if !path.exists() {
            return;
        }
        match fs::read_to_string(path)
            .map_err(|e| e.to_string())
            .and_then(|raw| crate::yaml::parse(&raw).map_err(|e| e.to_string()))
        {
            Ok(v) => *target = Some(v),
            Err(reason) => diags.push(TfDiagnostic {
                file: path.display().to_string(),
                reason,
            }),
        }
    };
    let try_json = |path: &Path, target: &mut Option<Value>, diags: &mut Vec<TfDiagnostic>| {
        if !path.exists() {
            return;
        }
        match fs::read_to_string(path)
            .map_err(|e| e.to_string())
            .and_then(|raw| serde_json::from_str::<Value>(&raw).map_err(|e| e.to_string()))
        {
            Ok(v) => *target = Some(v),
            Err(reason) => diags.push(TfDiagnostic {
                file: path.display().to_string(),
                reason,
            }),
        }
    };

    let agent = paths
        .agent_contract
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_AGENT_CONTRACT));
    try_yaml(&agent, &mut out.agent_contract, &mut out.diagnostics);
    let tm = paths
        .threat_model
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_THREAT_MODEL));
    try_yaml(&tm, &mut out.threat_model, &mut out.diagnostics);
    let policy = paths
        .policy
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_POLICY));
    try_yaml(&policy, &mut out.policy, &mut out.diagnostics);
    let actions = paths
        .actions
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_ACTIONS));
    try_yaml(&actions, &mut out.actions, &mut out.diagnostics);
    let pp = paths
        .proof_profile
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_PROOF_PROFILE));
    try_yaml(&pp, &mut out.proof_profile, &mut out.diagnostics);
    let cf = paths
        .conformance
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_CONFORMANCE));
    try_json(&cf, &mut out.conformance, &mut out.diagnostics);

    let cg = paths
        .codegen
        .clone()
        .unwrap_or_else(|| paths.root_dir.join(REL_CODEGEN));
    if cg.exists() {
        match fs::read_to_string(&cg) {
            Ok(raw) => out.codegen = Some(parse_tiny_toml(&raw)),
            Err(e) => out.diagnostics.push(TfDiagnostic {
                file: cg.display().to_string(),
                reason: e.to_string(),
            }),
        }
    }
    out
}

pub fn build_feature_gate(manifests: &TfManifests) -> TfFeatureGate {
    let mut gate = TfFeatureGate::default();
    gate.policy = manifests.policy.clone();
    if let Some(conf) = manifests.conformance.as_ref() {
        if let Some(arr) = conf.get("claimed_profiles").and_then(|v| v.as_array()) {
            gate.claimed_profiles = arr
                .iter()
                .filter_map(|v| v.as_str().map(str::to_string))
                .collect();
        }
    }
    if let Some(pp) = manifests.proof_profile.as_ref() {
        gate.default_proof_level = pp
            .get("default_proof_level")
            .and_then(|v| v.as_str())
            .map(str::to_string)
            .or_else(|| {
                pp.get("default_level")
                    .and_then(|v| v.as_str())
                    .map(str::to_string)
            });
        if let Some(actions) = pp.get("actions").and_then(|v| v.as_array()) {
            for a in actions {
                if let (Some(name), Some(level)) = (
                    a.get("name").and_then(|v| v.as_str()),
                    a.get("level").and_then(|v| v.as_str()),
                ) {
                    gate.per_action_proof_level
                        .insert(name.to_string(), level.to_string());
                }
            }
        }
        if let Some(anchors) = pp.get("anchors").and_then(|v| v.as_array()) {
            gate.anchors = anchors.clone();
        }
    }
    if let Some(ac) = manifests.agent_contract.as_ref() {
        if let Some(forbidden) = ac.get("forbidden").and_then(|v| v.as_array()) {
            for f in forbidden {
                if let Some(name) = f.get("action").and_then(|v| v.as_str()) {
                    gate.forbidden_actions.push(name.to_string());
                }
            }
        }
    }
    gate
}

fn parse_tiny_toml(raw: &str) -> HashMap<String, String> {
    let mut out = HashMap::new();
    for line in raw.lines() {
        let stripped = line.split('#').next().unwrap_or(line).trim();
        if stripped.is_empty() {
            continue;
        }
        if let Some((k, v)) = stripped.split_once('=') {
            let key = k.trim();
            let mut value = v.trim();
            if (value.starts_with('"') && value.ends_with('"'))
                || (value.starts_with('\'') && value.ends_with('\''))
            {
                value = &value[1..value.len() - 1];
            }
            out.insert(key.to_string(), value.to_string());
        }
    }
    out
}
